File: go/0039-combination-sum.go
package main

import "slices"

func combinationSum(candidates []int, target int) [][]int {
	var res [][]int
	var curr []int
	var currSum int

	combinationSum_backtrack(0, candidates, target, &currSum, &curr, &res)

	return res
}

func combinationSum_backtrack(index int, candidates []int, target int, currSum *int, curr *[]int, res *[][]int) {
	if *currSum == target {
		*res = append(*res, slices.Clone(*curr))
		return
	}
	if *currSum > target || (index < 0 || index >= len(candidates)) {
		return
	}

	// include same num again, duplicates are allowed
	*currSum += candidates[index]
	*curr = append(*curr, candidates[index])
	combinationSum_backtrack(index, candidates, target, currSum, curr, res)
	// backtrack, remove current num and include next one
	*currSum -= candidates[index]
	*curr = (*curr)[:len(*curr)-1]
	combinationSum_backtrack(index+1, candidates, target, currSum, curr, res)
}

File: java/0039-combination-sum.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        int currSum = 0;

        dfs(candidates, target, res, curr, currSum, 0);
        return res;
    }

    void dfs(int[] candidates, int target, List<List<Integer>> res, List<Integer> curr, int currSum, int i) {
        if (currSum == target) {
            res.add(List.copyOf(curr));
            return;
        }
        if (currSum > target || i < 0 || i >= candidates.length) {
            return;
        }

        curr.add(candidates[i]);
        currSum += candidates[i];
        dfs(candidates, target, res, curr, currSum, i);

        curr.remove(curr.size() - 1);
        currSum -= candidates[i];
        dfs(candidates, target, res, curr, currSum, i + 1);
    }
}